File: artifacts/sion-mlfb-decoder/src/knowledge/decoder.py
import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "sion_3ae5_knowledge.json"

with DATA.open(encoding="utf-8") as f:
    KB = json.load(f)

PRIMARY = {}
for item in KB["z_codes"]:
    pass

def normalize_code(code: str) -> str:
    """Normalize spaces/hyphens while preserving the -Z marker."""
    code = code.strip().upper()
    code = re.sub(r"\s+", "", code)
    return code

def split_input(raw: str):
    """
    Return base MLFB and additional order/Z codes.
    Examples:
      3AE5124-2AC90-6KN0-ZL1B+F30
      3AE5124-2AC90-6KN0-Z L1B F30
    """
    raw = normalize_code(raw)
    raw = raw.replace("+", " ")
    if "-Z" in raw:
        base, suffix = raw.split("-Z", 1)
        extras = [x for x in re.split(r"[\s,;]+", suffix) if x]
    else:
        base, extras = raw, []
    base = base.replace("-", "")
    return base, extras

def find_primary(article_number: str):
    # The catalog's primary-data tables identify the first 8 positions
    # as an article-number prefix such as 3AE5124-2.
    import csv
    path = ROOT / "data" / "primary_article_lookup.csv"
    with path.open(encoding="utf-8") as f:
        for row in csv.DictReader(f):
            if row["article_number"].replace("-", "") == article_number.replace("-", ""):
                return row
    return None

def decode_base(base: str):
    if len(base) != 16:
        raise ValueError(
            f"Expected a 16-position article number after removing separators; got {len(base)} characters."
        )
    if not base.startswith("3AE5"):
        raise ValueError("This starter decoder is for Siemens SION 3AE5 only.")

    pos = {str(i + 1): ch for i, ch in enumerate(base)}
    primary_article = f"{base[:7]}-{base[7]}"
    lookup = find_primary(primary_article)

    result = {
        "product": "Siemens SION vacuum circuit-breaker 3AE5",
        "raw_base": base,
        "positions": pos,
        "primary_lookup": lookup,
        "decoded": [],
        "warnings": [],
    }

    p = KB["position_codes"]

    result["decoded"].extend([
        {"position": 1, "value": pos["1"], "meaning": "Switching devices"},
        {"position": 2, "value": pos["2"], "meaning": "Circuit-breaker"},
        {"position": 3, "value": pos["3"], "meaning": "Circuit-breaker type series"},
        {"position": 4, "value": pos["4"], "meaning": "SION 3AE5 circuit-breaker version"},
    ])

    voltage_code = pos["5"]
    voltage_map = p["5"]["codes"]
    if voltage_code in voltage_map:
        result["decoded"].append({
            "position": 5,
            "value": voltage_code,
            "meaning": f"Rated voltage: {voltage_map[voltage_code]} kV"
        })
    else:
        result["warnings"].append(f"Unknown position-5 voltage code: {voltage_code}")

    if lookup:
        result["decoded"].extend([
            {"position": 6, "value": pos["6"],
             "meaning": f"Pole-center distance: {lookup['pole_center_distance_mm']} mm; "
                        f"vertical terminal distance: {lookup['vertical_distance_between_terminals_mm']} mm"},
            {"position": 7, "value": pos["7"],
             "meaning": f"Rated short-circuit breaking current: {lookup['rated_short_circuit_breaking_current_ka']} kA"},
            {"position": 8, "value": pos["8"],
             "meaning": f"Rated continuous current: {lookup['rated_continuous_current_a']} A"},
        ])
    else:
        result["warnings"].append(
            "Exact primary article number was not found in the catalog lookup table; "
            "positions 6–8 should not be guessed."
        )

    if pos["9"] in p["9"]["codes"]:
        result["decoded"].append({
            "position": 9, "value": pos["9"],
            "meaning": p["9"]["codes"][pos["9"]]
        })

    if pos["10"] in p["10"]["codes"]:
        result["decoded"].append({
            "position": 10, "value": pos["10"],
            "meaning": f"Closing solenoid: {p['10']['codes'][pos['10']]}"
        })

    if pos["11"] in p["11-12"]["standard_code_map"]:
        result["decoded"].append({
            "position": "11–12", "value": pos["11"] + pos["12"],
            "meaning": f"1st/2nd release operating voltage: {p['11-12']['standard_code_map'][pos['11']]}"
        })
    elif pos["11"] == "9":
        result["decoded"].append({
            "position": "11–12", "value": pos["11"] + pos["12"],
            "meaning": "Special release voltage; a descriptive order code is required."
        })

    if pos["13"] in p["13"]["codes"]:
        result["decoded"].append({
            "position": 13, "value": pos["13"],
            "meaning": p["13"]["codes"][pos["13"]]
        })

    if pos["14"] in p["14"]["codes"]:
        result["decoded"].append({
            "position": 14, "value": pos["14"],
            "meaning": f"Drive motor operating voltage: {p['14']['codes'][pos['14']]}"
        })

    if pos["15"] in p["15"]["codes"]:
        result["decoded"].append({
            "position": 15, "value": pos["15"],
            "meaning": p["15"]["codes"][pos["15"]]
        })

    if pos["16"] in p["16"]["codes"]:
        result["decoded"].append({
            "position": 16, "value": pos["16"],
            "meaning": p["16"]["codes"][pos["16"]]
        })

    return result

def decode(raw: str):
    base, extras = split_input(raw)
    result = decode_base(base)

    zmap = {x["code"]: x for x in KB["z_codes"]}
    result["order_codes"] = []

    for code in extras:
        if code in zmap:
            result["order_codes"].append(zmap[code])
        else:
            result["warnings"].append(f"Unknown/unloaded order code: {code}")

    codes = set(extras)

    # Source-derived compatibility checks.
    if {"A29", "A30"} <= codes:
        result["warnings"].append("INVALID: A29 and A30 are mutually exclusive.")
    if {"A47", "J60"} <= codes:
        result["warnings"].append("INVALID: A47 and J60 are mutually exclusive.")
    if "W88" in codes and "D93" not in codes:
        result["warnings"].append("INVALID/INCOMPLETE: W88 requires D93.")
    if "W89" in codes and "D93" not in codes:
        result["warnings"].append("INVALID/INCOMPLETE: W89 requires D93.")
    if ("M04" in codes or "M05" in codes) and not ({"W88", "W89"} & codes):
        result["warnings"].append("INVALID/INCOMPLETE: M04/M05 require W88 or W89.")
    if "S49" in codes and result["positions"]["13"] != "0":
        result["warnings"].append("INVALID: S49 is only possible for fixed mounting.")
    if any(c in codes for c in ["B01","B02","B03","B04","B05","B06","B07","B08","B09","B17"]) and result["positions"]["15"] != "X":
        result["warnings"].append("INVALID/INCOMPLETE: selected cable-harness options require position 15 = X.")

    # Resolve common special-order codes into human-readable meanings.
    result["special_order_codes"] = [
        x for x in result["order_codes"]
        if x.get("category") == "special order code"
    ]

    return result

if __name__ == "__main__":
    import sys
    if len(sys.argv) != 2:
        print("Usage: python src/decoder.py '3AE5124-2AC90-6KN0-ZL1B+F30'")
        raise SystemExit(1)

    result = decode(sys.argv[1])
    print(json.dumps(result, indent=2, ensure_ascii=False))

File: frontend/src/lib/decoder.ts
import knowledge from '../knowledge/sion_3ae5_knowledge.json';
import primaryCsv from '../knowledge/primary_article_lookup.csv?raw';

export type ZCode = {
  code: string;
  description: string;
  remarks?: string;
  category: string;
};

export type PositionResult = {
  position: string;
  value: string;
  label: string;
  meaning: string;
  sourcePage: string;
  unknown?: boolean;
};

export type PrimaryRow = {
  article_number: string;
  rated_voltage_kv: string;
  rated_short_circuit_breaking_current_ka: string;
  pole_center_distance_mm: string;
  vertical_distance_between_terminals_mm: string;
  rated_continuous_current_a: string;
};

export type DecodeResult = {
  raw: string;
  base: string;
  formattedBase: string;
  extras: string[];
  positions: Record<string, string>;
  positionResults: PositionResult[];
  primary: PrimaryRow | null;
  orderCodes: ZCode[];
  unknownOrderCodes: string[];
  warnings: string[];
  valid: boolean;
};

const KB = knowledge as any;
const zCodes = KB.z_codes as ZCode[];

const parseCsv = (csv: string): PrimaryRow[] => {
  const rows = csv.trim().split(/\r?\n/);
  const headers = rows.shift()?.split(',') ?? [];
  return rows.map((line) => {
    const values = line.split(',');
    return headers.reduce((row, header, index) => {
      row[header as keyof PrimaryRow] = values[index] ?? '';
      return row;
    }, {} as PrimaryRow);
  });
};

const primaryRows = parseCsv(primaryCsv);

export const catalog = zCodes;
export const primaryCatalog = primaryRows;
export const exampleCode = '3AE5124-2AC90-6KN0-ZL1B+F30';

export function normalizeCode(raw: string): string {
  return raw.trim().toUpperCase().replace(/\s+/g, '');
}

export function formatInput(raw: string): string {
  const compact = raw.toUpperCase().replace(/[^A-Z0-9]/g, '');
  if (!compact) return '';

  const base = compact.slice(0, 16);
  const suffixWithMarker = compact.slice(16);
  const hasExplicitMarker = suffixWithMarker.startsWith('Z');
  const suffix = hasExplicitMarker ? suffixWithMarker.slice(1) : suffixWithMarker;

  let formattedBase = base.slice(0, 7);
  if (base.length > 7) formattedBase += `-${base.slice(7, 12)}`;
  if (base.length > 12) formattedBase += `-${base.slice(12, 16)}`;

  if (!hasExplicitMarker && !suffix) return formattedBase;

  const formattedSuffix = (suffix.match(/.{1,3}/g) ?? []).join('+');
  return `${formattedBase}-Z${formattedSuffix}`;
}

export function splitInput(raw: string): { base: string; extras: string[] } {
  const normalized = normalizeCode(raw).replace(/\+/g, ' ');
  const markerIndex = normalized.indexOf('-Z');
  if (markerIndex < 0) {
    return { base: normalized.replace(/-/g, ''), extras: [] };
  }
  const base = normalized.slice(0, markerIndex).replace(/-/g, '');
  let suffix = normalized.slice(markerIndex + 2).replace(/-/g, ' ');
  let extras = suffix.split(/[\s,;]+/).filter(Boolean);
  if (extras.length === 1 && extras[0].length > 3 && extras[0].length % 3 === 0) {
    extras = extras[0].match(/.{3}/g) ?? extras;
  }
  return { base, extras };
}

const primaryFor = (base: string): PrimaryRow | null => {
  const article = `${base.slice(0, 7)}-${base.slice(7, 8)}`;
  return primaryRows.find((row) => row.article_number.replace(/-/g, '') === article.replace(/-/g, '')) ?? null;
};

const sourcePageFor = (position: string): string => {
  if (['1', '2', '3', '4'].includes(position)) return 'p. 14';
  if (['5', '6', '7', '8'].includes(position)) return 'pp. 16–24';
  if (position === '9') return 'p. 25';
  if (position === '10') return 'p. 26';
  if (position === '11–12') return 'p. 27';
  if (position === '13') return 'p. 28';
  if (position === '14') return 'p. 29';
  if (position === '15') return 'p. 30';
  return 'p. 31';
};

const lookup = (position: string, value: string, label: string, meaning: string): PositionResult => ({
  position, value, label, meaning, sourcePage: sourcePageFor(position),
});

function knownOrUnknown(
  position: string,
  value: string,
  label: string,
  mapping: Record<string, string | number> | undefined,
  prefix = '',
): PositionResult {
  if (mapping && value in mapping) return lookup(position, value, label, `${prefix}${mapping[value]}`);
  return { ...lookup(position, value, label, `Unknown code: ${value}`), unknown: true };
}

export function decode(raw: string): DecodeResult {
  const { base, extras } = splitInput(raw);
  if (!base) throw new Error('Enter an article number to begin.');
  if (base.length !== 16) {
    throw new Error(`Expected 16 positions after removing separators; received ${base.length}.`);
  }
  if (!base.startsWith('3AE5')) throw new Error('This decoder is limited to Siemens SION 3AE5 article numbers.');

  const positions = Object.fromEntries([...base].map((char, index) => [String(index + 1), char]));
  const primary = primaryFor(base);
  const p = KB.position_codes;
  const positionResults: PositionResult[] = [
    lookup('1', positions['1'], 'Superior group', 'Switching devices'),
    lookup('2', positions['2'], 'Main group', 'Circuit-breaker'),
    lookup('3', positions['3'], 'Subgroup', 'Circuit-breaker type series'),
    lookup('4', positions['4'], 'Circuit-breaker version', 'SION 3AE5 circuit-breaker version'),
    knownOrUnknown('5', positions['5'], 'Rated voltage', p['5'].codes, 'Rated voltage: '),
  ];

  if (primary) {
    positionResults.push(
      lookup('6', positions['6'], 'Pole-center / terminal distance', `${primary.pole_center_distance_mm} mm pole-center; ${primary.vertical_distance_between_terminals_mm} mm terminal distance`),
      lookup('7', positions['7'], 'Short-circuit breaking current', `${primary.rated_short_circuit_breaking_current_ka} kA`),
      lookup('8', positions['8'], 'Continuous current', `${primary.rated_continuous_current_a} A`),
    );
  } else {
    ['6', '7', '8'].forEach((position) => {
      positionResults.push({
        ...lookup(position, positions[position], position === '6' ? 'Primary dimensions' : position === '7' ? 'Short-circuit current' : 'Continuous current', `Unknown code: ${positions[position]} — exact primary article not in catalog`),
        unknown: true,
      });
    });
  }

  positionResults.push(
    knownOrUnknown('9', positions['9'], 'Release combination', p['9'].codes),
    knownOrUnknown('10', positions['10'], 'Closing solenoid', p['10'].codes, 'Closing solenoid: '),
  );
  if (positions['11'] in p['11-12'].standard_code_map) {
    positionResults.push(lookup('11–12', `${positions['11']}${positions['12']}`, '1st / 2nd release voltage', p['11-12'].standard_code_map[positions['11']]));
  } else if (positions['11'] === '9') {
    positionResults.push(lookup('11–12', `${positions['11']}${positions['12']}`, '1st / 2nd release voltage', 'Special release voltage; a descriptive order code is required.'));
  } else {
    positionResults.push({ ...lookup('11–12', `${positions['11']}${positions['12']}`, '1st / 2nd release voltage', `Unknown code: ${positions['11']}${positions['12']}`), unknown: true });
  }
  positionResults.push(
    knownOrUnknown('13', positions['13'], 'Installation options', p['13'].codes),
    knownOrUnknown('14', positions['14'], 'Drive motor voltage', p['14'].codes, 'Drive motor: '),
    knownOrUnknown('15', positions['15'], 'Low-voltage interface', p['15'].codes),
    knownOrUnknown('16', positions['16'], 'Language', p['16'].codes),
  );

  const zByCode = new Map(zCodes.map((item) => [item.code, item]));
  const orderCodes = extras.filter((code) => zByCode.has(code)).map((code) => zByCode.get(code) as ZCode);
  const unknownOrderCodes = extras.filter((code) => !zByCode.has(code));
  const warnings: string[] = [];
  if (!primary) warnings.push('Exact primary article number was not found in the catalog lookup table; positions 6–8 are not guessed.');
  if (unknownOrderCodes.length) unknownOrderCodes.forEach((code) => warnings.push(`Unknown order code: ${code}`));
  const codeSet = new Set(extras);
  if (codeSet.has('A29') && codeSet.has('A30')) warnings.push('INVALID: A29 and A30 are mutually exclusive.');
  if (codeSet.has('A47') && codeSet.has('J60')) warnings.push('INVALID: A47 and J60 are mutually exclusive.');
  if (codeSet.has('W88') && !codeSet.has('D93')) warnings.push('INVALID/INCOMPLETE: W88 requires D93.');
  if (codeSet.has('W89') && !codeSet.has('D93')) warnings.push('INVALID/INCOMPLETE: W89 requires D93.');
  if ((codeSet.has('M04') || codeSet.has('M05')) && !codeSet.has('W88') && !codeSet.has('W89')) warnings.push('INVALID/INCOMPLETE: M04/M05 require W88 or W89.');
  if ((codeSet.has('M04') || codeSet.has('M05')) && (codeSet.has('A29') || codeSet.has('A30'))) warnings.push('INVALID: M04/M05 are not compatible with A29 or A30.');
  if (codeSet.has('S49') && positions['13'] !== '0') warnings.push('INVALID: S49 is only possible for fixed mounting.');
  const harnessCodes = ['B01', 'B02', 'B03', 'B04', 'B05', 'B06', 'B07', 'B08', 'B09', 'B17'];
  if (harnessCodes.some((code) => codeSet.has(code)) && positions['15'] !== 'X') warnings.push('INVALID/INCOMPLETE: selected cable-harness options require position 15 = X.');

  return {
    raw,
    base,
    formattedBase: `${base.slice(0, 7)}-${base.slice(7, 8)}${base.slice(8, 12)}-${base.slice(12)}`,
    extras,
    positions,
    positionResults,
    primary,
    orderCodes,
    unknownOrderCodes,
    warnings,
    valid: !warnings.some((warning) => warning.startsWith('INVALID')),
  };
}

export function humanDescription(result: DecodeResult): string {
  if (!result.primary) return 'A Siemens SION 3AE5 configuration with an unlisted primary article. Electrical dimensions and ratings are intentionally not inferred.';
  const p = result.primary;
  const extras = result.orderCodes.map((item) => `${item.code} (${item.description.toLowerCase()})`).join('; ');
  const tail = extras ? ` Additional order codes: ${extras}.` : '';
  return `Siemens SION 3AE5 vacuum circuit-breaker, ${p.rated_voltage_kv} kV, ${p.rated_short_circuit_breaking_current_ka} kA, ${p.rated_continuous_current_a} A, ${p.pole_center_distance_mm} mm pole-center distance and ${p.vertical_distance_between_terminals_mm} mm vertical terminal distance.${tail}`;
}

export function sourceFor(result: DecodeResult): { label: string; page: string }[] {
  const sources = new Map<string, string>();
  result.positionResults.forEach((item) => sources.set(item.label, item.sourcePage));
  if (result.orderCodes.length) sources.set('Additional order codes', 'pp. 32–33');
  return [...sources].map(([label, page]) => ({ label, page }));
}